lib/src/package_analysis/lower_bound_constraint_analysis.dart

```diff
@@ -0,0 +1,630 @@
+      throw AnalysisException('Could not determine library of parentElement.');
+    }
+    final tryPackageName =
+        packageFromLibraryUri(parentElement.library!.identifier);
+    // if the defining package isn't a HostedDependency of the target, then
+    // this symbol cannot be analyzed
+    if (tryPackageName == null ||
+        !context.dependencies.keys.contains(tryPackageName)) {
+      throw AnalysisException(
+          'The defining package is not a HostedDependency of the target package.');
+    }
+    packageName = tryPackageName;
+
+    if (!dependencySummaries.keys.contains(packageName)) {
+      throw AnalysisException(
+          'No summary was found for the defining package $packageName.');
+    }
+    dependencyShape = dependencySummaries[packageName]!;
+  }
+
+  /// Populate the various properties of this visitor based on information from
+  /// this symbol, throwing an [AnalysisException] if the symbol does not need
+  /// to be/cannot be analyzed.
+  void processIdentifier(SimpleIdentifier identifier) {
+    span = SourceFile.fromString(
+      context.readFile(context.uriToPath(currentUri)!),
+      url: currentUri,
+    ).span(identifier.offset, identifier.end);
+
+    elementId = element.id;
+    // if the element is a getter or a setter, pull out just the variable name
+    symbolName = element is PropertyAccessorElement
+        ? (element as PropertyAccessorElement).variable.name
+        : element.name!;
+
+    // if we have seen this symbol before, we need to do no further analysis
+    if (issues.containsKey(elementId)) {
+      // if we have seen this element before and there is an issue with it,
+      // add this usage/invocation to its list of references
+      issues[elementId]?.references.add(span);
+      throw AnalysisException(
+          'The definition of this Element has already been visited.');
+    }
+
+    const supportedKinds = [
+      ElementKind.FUNCTION,
+      ElementKind.METHOD,
+      ElementKind.GETTER,
+      ElementKind.SETTER,
+    ];
+    if (!supportedKinds.contains(element.kind)) {
+      // prior checks should have filtered out any invocations other than these
+      throw StateError('Unexpected ElementKind ${element.kind}.');
+    }
+  }
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    // this must be able to handle cases of a library split into multiple files
+    // with the part keyword
+    currentUri = node.declaredElement!.source.uri;
+    super.visitCompilationUnit(node);
+  }
+
+  // TODO: do we need visitPropertyAccess too?
+  // Invocations of getters and setters are represented by either PrefixedIdentifier or PropertyAccess nodes.
+  // PropertyAccess: The access of a property of an object. Note, however, that
+  // accesses to properties of objects can also be represented as
+  // PrefixedIdentifier nodes in cases where the target is also a simple identifier.
+  @override
+  void visitPrefixedIdentifier(PrefixedIdentifier node) {
```

Comment:
do we need visitPropertyAccess too?
https://pub.dev/documentation/analyzer/latest/dart_ast_ast/PropertyAccess-class.html
> The access of a property of an object.
> Note, however, that accesses to properties of objects can also be represented as [PrefixedIdentifier](https://pub.dev/documentation/analyzer/latest/dart_ast_ast/PrefixedIdentifier-class.html) nodes in cases where the target is also a simple identifier.

What is [Expression](https://pub.dev/documentation/analyzer/latest/dart_ast_ast/Expression-class.html) and should we consider cases where it is the `node.target`?